flaky test TEST_CLI

**Observed behavior**
--- FAIL: Test_CLI (134.71s)
    --- FAIL: Test_CLI/deploy_testdata/corerp-kubernetes-cli.bicep (105.03s)
        --- PASS: Test_CLI/deploy_testdata/corerp-kubernetes-cli.bicep/Validate_rad_application_show (2.74s)
        --- PASS: Test_CLI/deploy_testdata/corerp-kubernetes-cli.bicep/Validate_rad_resource_list (34.61s)
        --- PASS: Test_CLI/deploy_testdata/corerp-kubernetes-cli.bicep/Validate_rad_resource_show (3.38s)
        --- PASS: Test_CLI/deploy_testdata/corerp-kubernetes-cli.bicep/Validate_rad_resoure_logs_containers (11.21s)
        --- FAIL: Test_CLI/deploy_testdata/corerp-kubernetes-cli.bicep/Validate_rad_resource_expose_Container (9.01s)

https://dev.azure.com/azure-octo/Incubations/_build/results?buildId=34464&view=logs&j=4b2e8509-34d5-53da-a1f6-67bf3fd840d5&t=e9705480-daaf-589d-7a3b-4dfdd6fc3817

**Desired behavior**
<!--What you're expecting to happen-->

**OS/rad version**
<!--What operating system and rad cli version are you running?-->

<!-- PASTE OUTPUT OF rad version -->

**Screenshots**
<!--If applicable, add screenshots to help explain your problem-->

**Additional context**
<!--Add any other context about the problem here-->

AB#4196

--- a/go.mod
+++ b/go.mod
@@ -94,6 +94,7 @@ require (
 
 require (
 	github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da // indirect
+	github.com/hashicorp/go-retryablehttp v0.7.5 // indirect
 	github.com/stretchr/objx v0.5.1 // indirect
 	github.com/tidwall/gjson v1.14.0 // indirect
 	github.com/tidwall/match v1.1.1 // indirect

--- a/test/functional/shared/cli/cli_test.go
+++ b/test/functional/shared/cli/cli_test.go
@@ -260,7 +260,7 @@ func callHealthEndpointOnLocalPort(t *testing.T, retries int, port int) {
 	retryClient := retryablehttp.NewClient()
 	retryClient.RetryMax = retries
 	retryClient.RetryWaitMin = 5 * time.Second
-	retryClient.RetryWaitMax = 60 * time.Second
+	retryClient.RetryWaitMax = 20 * time.Second
 	retryClient.Backoff = retryablehttp.LinearJitterBackoff
 	retryClient.RequestLogHook = func(_ retryablehttp.Logger, req *http.Request, retry int) {
 		t.Logf("retry calling healthz endpoint %s, retry: %d", healthzURL, retry)

--- a/test/functional/shared/cli/cli_test.go
+++ b/test/functional/shared/cli/cli_test.go
@@ -259,12 +259,12 @@ func callHealthEndpointOnLocalPort(t *testing.T, retries int, port int) {
 
 	retryClient := retryablehttp.NewClient()
 	retryClient.RetryMax = retries
-	retryClient.RequestLogHook = func(_ retryablehttp.Logger, req *http.Request, retry int) {
-		t.Logf("retry calling healthz endpoint %s, retry: %d", healthzURL, retry)
-	}
 	retryClient.RetryWaitMin = 5 * time.Second
 	retryClient.RetryWaitMax = 60 * time.Second
 	retryClient.Backoff = retryablehttp.LinearJitterBackoff
+	retryClient.RequestLogHook = func(_ retryablehttp.Logger, req *http.Request, retry int) {
+		t.Logf("retry calling healthz endpoint %s, retry: %d", healthzURL, retry)
+	}
 
 	resp, err := retryClient.Get(healthzURL)
 	require.NoError(t, err, "failed to get connect to resource after %d retries", retries)

--- a/test/functional/shared/cli/cli_test.go
+++ b/test/functional/shared/cli/cli_test.go
@@ -33,6 +33,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/hashicorp/go-retryablehttp"
 	"github.com/radius-project/radius/pkg/cli/bicep"
 	"github.com/radius-project/radius/pkg/cli/clients"
 	"github.com/radius-project/radius/pkg/cli/cmd/radinit"
@@ -254,40 +255,24 @@ func verifyCLIBasics(ctx context.Context, t *testing.T, test shared.RPTest) {
 // callHealthEndpointOnLocalPort calls the magpie health endpoint '/healthz' with retries. It will fail the
 // test if the exceed the number of retries without success.
 func callHealthEndpointOnLocalPort(t *testing.T, retries int, port int) {
-	for i := 0; i < retries; i++ {
-		url := fmt.Sprintf("http://localhost:%d/healthz", port)
-		t.Logf("making request to %s", url)
-		response, err := http.Get(url)
-		if err != nil {
-			if i == retries-1 {
-				// last retry failed, report failure
-				require.NoError(t, err, "failed to get connect to resource after %d retries", retries)
-			}
-			t.Logf("got error %s", err.Error())
-			time.Sleep(5 * time.Second)
-			continue
-		}
-		if response.Body != nil {
-			defer response.Body.Close()
-		}
+	healthzURL := fmt.Sprintf("http://localhost:%d/healthz", port)
 
-		if response.StatusCode > 299 || response.StatusCode < 200 {
-			if i == retries-1 {
-				// last retry failed, report failure
-				require.NoError(t, err, "status code was a bad response after %d retries %d", retries, response.StatusCode)
-			}
-			t.Logf("got status %d", response.StatusCode)
-			time.Sleep(1 * time.Second)
-			continue
-		}
-
-		defer response.Body.Close()
-		content, err := io.ReadAll(response.Body)
-		require.NoError(t, err)
-
-		t.Logf("[response] %s", string(content))
-		return
+	retryClient := retryablehttp.NewClient()
+	retryClient.RetryMax = retries
+	retryClient.RequestLogHook = func(_ retryablehttp.Logger, req *http.Request, retry int) {
+		t.Logf("retry calling healthz endpoint %s, retry: %d", healthzURL, retry)
 	}
+	retryClient.RetryWaitMin = 5 * time.Second
+	retryClient.RetryWaitMax = 60 * time.Second
+	retryClient.Backoff = retryablehttp.LinearJitterBackoff
+
+	resp, err := retryClient.Get(healthzURL)
+	require.NoError(t, err, "failed to get connect to resource after %d retries", retries)
+	defer resp.Body.Close()
+	content, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+
+	t.Logf("[response] %s", string(content))
 }
 
 func Test_Run_Logger(t *testing.T) {

--- a/test/functional/shared/cli/cli_test.go
+++ b/test/functional/shared/cli/cli_test.go
@@ -24,7 +24,6 @@ import (
 	"io"
 	"math/rand"
 	"net"
-	"net/http"
 	"os"
 	"path/filepath"
 	"regexp"
@@ -33,6 +32,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/hashicorp/go-retryablehttp"
 	"github.com/radius-project/radius/pkg/cli/bicep"
 	"github.com/radius-project/radius/pkg/cli/clients"
 	"github.com/radius-project/radius/pkg/cli/cmd/radinit"
@@ -228,66 +228,37 @@ func verifyCLIBasics(ctx context.Context, t *testing.T, test shared.RPTest) {
 	})
 
 	t.Run("Validate rad resource expose Container", func(t *testing.T) {
-		t.Skip("https://github.com/radius-project/radius/issues/3232")
 		port, err := GetAvailablePort()
 		require.NoError(t, err)
 
 		// We open a local port-forward and then make a request to it.
 		child, cancel := context.WithCancel(ctx)
+		t.Cleanup(cancel)
 
-		done := make(chan error)
+		errCh := make(chan error)
 		go func() {
 			_, err = cli.ResourceExpose(child, appName, containerName, port, 3000)
-			done <- err
+			errCh <- err
 		}()
 
 		callHealthEndpointOnLocalPort(t, retries, port)
-
-		cancel()
-		err = <-done
-
-		// The error should be due to cancellation (we can canceled the command).
-		require.Equal(t, context.Canceled, err)
+		require.NoError(t, <-errCh, "failed to expose endpoint from container")
 	})
 }
 
 // callHealthEndpointOnLocalPort calls the magpie health endpoint '/healthz' with retries. It will fail the
 // test if the exceed the number of retries without success.
 func callHealthEndpointOnLocalPort(t *testing.T, retries int, port int) {
-	for i := 0; i < retries; i++ {
-		url := fmt.Sprintf("http://localhost:%d/healthz", port)
-		t.Logf("making request to %s", url)
-		response, err := http.Get(url)
-		if err != nil {
-			if i == retries-1 {
-				// last retry failed, report failure
-				require.NoError(t, err, "failed to get connect to resource after %d retries", retries)
-			}
-			t.Logf("got error %s", err.Error())
-			time.Sleep(1 * time.Second)
-			continue
-		}
-		if response.Body != nil {
-			defer response.Body.Close()
-		}
+	retryClient := retryablehttp.NewClient()
+	retryClient.RetryMax = retries
 
-		if response.StatusCode > 299 || response.StatusCode < 200 {
-			if i == retries-1 {
-				// last retry failed, report failure
-				require.NoError(t, err, "status code was a bad response after %d retries %d", retries, response.StatusCode)
-			}
-			t.Logf("got status %d", response.StatusCode)
-			time.Sleep(1 * time.Second)
-			continue
-		}
+	resp, err := retryClient.Get(fmt.Sprintf("http://localhost:%d/healthz", port))
+	require.NoError(t, err, "failed to get connect to resource after %d retries", retries)
+	defer resp.Body.Close()
+	content, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
 
-		defer response.Body.Close()
-		content, err := io.ReadAll(response.Body)
-		require.NoError(t, err)
-
-		t.Logf("[response] %s", string(content))
-		return
-	}
+	t.Logf("[response] %s", string(content))
 }
 
 func Test_Run_Logger(t *testing.T) {

--- a/test/functional/shared/cli/cli_test.go
+++ b/test/functional/shared/cli/cli_test.go
@@ -233,7 +233,6 @@ func verifyCLIBasics(ctx context.Context, t *testing.T, test shared.RPTest) {
 
 		// We open a local port-forward and then make a request to it.
 		child, cancel := context.WithCancel(ctx)
-		t.Cleanup(cancel)
 
 		errCh := make(chan error)
 		go func() {
@@ -242,6 +241,7 @@ func verifyCLIBasics(ctx context.Context, t *testing.T, test shared.RPTest) {
 		}()
 
 		callHealthEndpointOnLocalPort(t, retries, port)
+		cancel()
 		require.NoError(t, <-errCh, "failed to expose endpoint from container")
 	})
 }

--- a/test/functional/shared/cli/cli_test.go
+++ b/test/functional/shared/cli/cli_test.go
@@ -24,6 +24,7 @@ import (
 	"io"
 	"math/rand"
 	"net"
+	"net/http"
 	"os"
 	"path/filepath"
 	"regexp"
@@ -249,10 +250,15 @@ func verifyCLIBasics(ctx context.Context, t *testing.T, test shared.RPTest) {
 // callHealthEndpointOnLocalPort calls the magpie health endpoint '/healthz' with retries. It will fail the
 // test if the exceed the number of retries without success.
 func callHealthEndpointOnLocalPort(t *testing.T, retries int, port int) {
-	retryClient := retryablehttp.NewClient()
-	retryClient.RetryMax = retries
+	client := retryablehttp.NewClient()
+	client.RetryMax = retries
+	client.RequestLogHook = func(_ retryablehttp.Logger, _ *http.Request, retry int) {
+		t.Logf("failed to get connect to resource - %d", retry)
+	}
+	client.RetryWaitMin = 2 * time.Second
+	client.Backoff = retryablehttp.LinearJitterBackoff
 
-	resp, err := retryClient.Get(fmt.Sprintf("http://localhost:%d/healthz", port))
+	resp, err := client.Get(fmt.Sprintf("http://localhost:%d/healthz", port))
 	require.NoError(t, err, "failed to get connect to resource after %d retries", retries)
 	defer resp.Body.Close()
 	content, err := io.ReadAll(resp.Body)

--- a/test/functional/shared/cli/cli_test.go
+++ b/test/functional/shared/cli/cli_test.go
@@ -237,8 +237,8 @@ func verifyCLIBasics(ctx context.Context, t *testing.T, test shared.RPTest) {
 		done := make(chan error)
 		go func() {
 			output, err := cli.ResourceExpose(child, appName, containerName, port, 3000)
-			done <- err
 			t.Logf("ResourceExpose - output: %s", output)
+			done <- err
 		}()
 
 		callHealthEndpointOnLocalPort(t, retries, port)
@@ -264,7 +264,7 @@ func callHealthEndpointOnLocalPort(t *testing.T, retries int, port int) {
 				require.NoError(t, err, "failed to get connect to resource after %d retries", retries)
 			}
 			t.Logf("got error %s", err.Error())
-			time.Sleep(1 * time.Second)
+			time.Sleep(5 * time.Second)
 			continue
 		}
 		if response.Body != nil {

--- a/test/functional/shared/cli/cli_test.go
+++ b/test/functional/shared/cli/cli_test.go
@@ -236,8 +236,9 @@ func verifyCLIBasics(ctx context.Context, t *testing.T, test shared.RPTest) {
 
 		done := make(chan error)
 		go func() {
-			_, err = cli.ResourceExpose(child, appName, containerName, port, 3000)
+			output, err := cli.ResourceExpose(child, appName, containerName, port, 3000)
 			done <- err
+			t.Logf("ResourceExpose - output: %s", output)
 		}()
 
 		callHealthEndpointOnLocalPort(t, retries, port)

--- a/test/functional/shared/cli/cli_test.go
+++ b/test/functional/shared/cli/cli_test.go
@@ -33,7 +33,6 @@ import (
 	"testing"
 	"time"
 
-	"github.com/hashicorp/go-retryablehttp"
 	"github.com/radius-project/radius/pkg/cli/bicep"
 	"github.com/radius-project/radius/pkg/cli/clients"
 	"github.com/radius-project/radius/pkg/cli/cmd/radinit"
@@ -235,36 +234,59 @@ func verifyCLIBasics(ctx context.Context, t *testing.T, test shared.RPTest) {
 		// We open a local port-forward and then make a request to it.
 		child, cancel := context.WithCancel(ctx)
 
-		errCh := make(chan error)
+		done := make(chan error)
 		go func() {
 			_, err = cli.ResourceExpose(child, appName, containerName, port, 3000)
-			errCh <- err
+			done <- err
 		}()
 
 		callHealthEndpointOnLocalPort(t, retries, port)
+
 		cancel()
-		require.NoError(t, <-errCh, "failed to expose endpoint from container")
+		err = <-done
+
+		// The error should be due to cancellation (we can canceled the command).
+		require.Equal(t, context.Canceled, err)
 	})
 }
 
 // callHealthEndpointOnLocalPort calls the magpie health endpoint '/healthz' with retries. It will fail the
 // test if the exceed the number of retries without success.
 func callHealthEndpointOnLocalPort(t *testing.T, retries int, port int) {
-	client := retryablehttp.NewClient()
-	client.RetryMax = retries
-	client.RequestLogHook = func(_ retryablehttp.Logger, _ *http.Request, retry int) {
-		t.Logf("failed to get connect to resource - %d", retry)
+	for i := 0; i < retries; i++ {
+		url := fmt.Sprintf("http://localhost:%d/healthz", port)
+		t.Logf("making request to %s", url)
+		response, err := http.Get(url)
+		if err != nil {
+			if i == retries-1 {
+				// last retry failed, report failure
+				require.NoError(t, err, "failed to get connect to resource after %d retries", retries)
+			}
+			t.Logf("got error %s", err.Error())
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if response.Body != nil {
+			defer response.Body.Close()
+		}
+
+		if response.StatusCode > 299 || response.StatusCode < 200 {
+			if i == retries-1 {
+				// last retry failed, report failure
+				require.NoError(t, err, "status code was a bad response after %d retries %d", retries, response.StatusCode)
+			}
+			t.Logf("got status %d", response.StatusCode)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		defer response.Body.Close()
+		content, err := io.ReadAll(response.Body)
+		require.NoError(t, err)
+
+		t.Logf("[response] %s", string(content))
+		return
 	}
-	client.RetryWaitMin = 2 * time.Second
-	client.Backoff = retryablehttp.LinearJitterBackoff
-
-	resp, err := client.Get(fmt.Sprintf("http://localhost:%d/healthz", port))
-	require.NoError(t, err, "failed to get connect to resource after %d retries", retries)
-	defer resp.Body.Close()
-	content, err := io.ReadAll(resp.Body)
-	require.NoError(t, err)
-
-	t.Logf("[response] %s", string(content))
 }
 
 func Test_Run_Logger(t *testing.T) {

